From 06a15e993f3a1c57d70eeba91118c99866fb9b93 Mon Sep 17 00:00:00 2001
From: Michal Suchanek <msuchanek@suse.de>
Date: Mon, 27 Jul 2026 13:43:42 +0200
Subject: [PATCH] kABI: Fix ABI for "powerpc: Fix exit_flags field placement in
 pt_regs for ptrace"

References: bsc#1272455

Problem: pt_regs size increased which results in userspace passing too
small buffers for pt_regs.

Preserve kernel ABI by keeping the pt_regs padding that was previously added.

Preserve userspace ABI by copying only pt_regs part without padding.

Move new thread_info flags to existing flags field.

Signed-off-by: Michal Suchanek <msuchanek@suse.de>
Reviewed-by: Jan Kara <jack@suse.cz>
Reviewed-by: Takashi Iwai <tiwai@suse.de>
Reviewed-by: Petr Pavlu <petr.pavlu@suse.com>
---
 arch/powerpc/include/asm/ptrace.h      |    8 ++++++++
 arch/powerpc/include/asm/thread_info.h |   19 ++++++++++++++++++-
 arch/powerpc/include/uapi/asm/ptrace.h |    7 +++++++
 arch/powerpc/kernel/interrupt.c        |   19 ++++++++++---------
 arch/powerpc/kernel/ptrace/ptrace.c    |   10 ++++++----
 arch/powerpc/kernel/signal.c           |    2 +-
 6 files changed, 50 insertions(+), 15 deletions(-)

diff --git a/arch/powerpc/include/asm/ptrace.h b/arch/powerpc/include/asm/ptrace.h
index d53c4dd4d8b67..6fe6d31494a73 100644
--- a/arch/powerpc/include/asm/ptrace.h
+++ b/arch/powerpc/include/asm/ptrace.h
@@ -53,6 +53,14 @@ struct pt_regs
 				unsigned long esr;
 			};
 			unsigned long result;
+#ifndef __GENKSYMS__
+			unsigned long __pt_regs_pad[4];
+#else
+			unsigned long exit_flags;
+			unsigned long entry_flags;
+			/* Maintain 16 byte interrupt stack alignment */
+			unsigned long __pt_regs_pad[2];
+#endif
 		};
 	};
 #if defined(CONFIG_PPC64) || defined(CONFIG_PPC_KUAP)
diff --git a/arch/powerpc/include/asm/thread_info.h b/arch/powerpc/include/asm/thread_info.h
--- a/arch/powerpc/include/asm/thread_info.h
+++ b/arch/powerpc/include/asm/thread_info.h
@@ -57,7 +57,6 @@ struct thread_info {
 #ifdef CONFIG_SMP
 	unsigned int	cpu;
 #endif
-	unsigned long	exit_flags;		/* Exit Flags for entry/exit */
 	unsigned long	syscall_work;		/* SYSCALL_WORK_ flags */
 	unsigned long	local_flags;		/* private flags for thread */
 #ifdef CONFIG_LIVEPATCH_64
@@ -164,9 +163,17 @@ void arch_setup_new_exec(void);
 #define _TLF_SLEEPING		(1 << TLF_SLEEPING)
 #define _TLF_LAZY_MMU		(1 << TLF_LAZY_MMU)
 #define _TLF_RUNLATCH		(1 << TLF_RUNLATCH)
+/* Upstream these syscall flags are in a different field that cannot be added without breaking ABI */
+#define _TLF_SYSCALL_EXIT_RESTOREALL	BIT(29)
 
 #ifndef __ASSEMBLER__
 
+static inline void set_thread_local_flags(unsigned int flags)
+{
+	struct thread_info *ti = current_thread_info();
+	ti->local_flags |= flags;
+}
+
 static inline void clear_thread_local_flags(unsigned int flags)
 {
 	struct thread_info *ti = current_thread_info();
@@ -179,6 +186,16 @@ static inline bool test_thread_local_fla
 	return (ti->local_flags & flags) != 0;
 }
 
+static inline bool test_and_clear_thread_local_flags(unsigned int flags)
+{
+	struct thread_info *ti = current_thread_info();
+	bool ret = (ti->local_flags & flags) != 0;
+
+	ti->local_flags &= ~flags;
+
+	return ret;
+}
+
 #ifdef CONFIG_COMPAT
 #define is_32bit_task()	(test_thread_flag(TIF_32BIT))
 #define is_tsk_32bit_task(tsk)	(test_tsk_thread_flag(tsk, TIF_32BIT))
diff --git a/arch/powerpc/include/uapi/asm/ptrace.h b/arch/powerpc/include/uapi/asm/ptrace.h
index 01e630149d48e..626dd531cc553 100644
--- a/arch/powerpc/include/uapi/asm/ptrace.h
+++ b/arch/powerpc/include/uapi/asm/ptrace.h
@@ -55,6 +55,13 @@ struct pt_regs
 	unsigned long dar;		/* Fault registers */
 	unsigned long dsisr;		/* on 4xx/Book-E used for ESR */
 	unsigned long result;		/* Result of a system call */
+#ifndef __GENKSYMS__
+	unsigned long __pt_regs_pad[4];
+#else
+	unsigned long exit_flags;
+	unsigned long entry_flags;
+	unsigned long __pt_regs_pad[2];
+#endif
 };
 
 #endif /* __ASSEMBLER__ */
diff --git a/arch/powerpc/kernel/interrupt.c b/arch/powerpc/kernel/interrupt.c
--- a/arch/powerpc/kernel/interrupt.c
+++ b/arch/powerpc/kernel/interrupt.c
@@ -97,7 +97,7 @@ notrace unsigned long syscall_exit_prepare(unsigned long r3,
 	regs->result = r3;
 
 	/* Clear exit_flags so only flags set during this exit are visible */
-	current->thread_info.exit_flags = 0;
+	clear_thread_local_flags(_TLF_SYSCALL_EXIT_RESTOREALL);
 
 	ti_flags = read_thread_flags();
 	if (unlikely(r3 >= (unsigned long)-MAX_ERRNO) && is_not_scv) {
@@ -134,7 +134,8 @@ notrace unsigned long syscall_exit_prepare(unsigned long r3,
 
 	/* Restore user access locks last */
 	kuap_user_restore(regs);
-	ret |= current->thread_info.exit_flags;
+	if (test_thread_local_flags(_TLF_SYSCALL_EXIT_RESTOREALL))
+		ret |= _TIF_RESTOREALL;
 #ifdef CONFIG_PPC64
 	regs->exit_result = ret;
 #endif
@@ -145,8 +146,6 @@ notrace unsigned long syscall_exit_prepare(unsigned long r3,
 #ifdef CONFIG_PPC64
 notrace unsigned long syscall_exit_restart(unsigned long r3, struct pt_regs *regs)
 {
-	unsigned long ret;
-
 	/*
 	 * This is called when detecting a soft-pending interrupt as well as
 	 * an alternate-return interrupt. So we can't just have the alternate
@@ -171,9 +170,8 @@ notrace unsigned long syscall_exit_restart(unsigned long r3, struct pt_regs *reg
 	}
 
 	kuap_user_restore(regs);
-	ret = current_thread_info()->exit_flags & _TIF_RESTOREALL;
-	current_thread_info()->exit_flags &= ~_TIF_RESTOREALL;
-	regs->exit_result |= ret;
+	if (test_and_clear_thread_local_flags(_TLF_SYSCALL_EXIT_RESTOREALL))
+		regs->exit_result |= _TIF_RESTOREALL;
 
 	return regs->exit_result;
 }
@@ -181,7 +181,7 @@ notrace unsigned long syscall_exit_restart(unsigned long r3, struct pt_regs *reg
 
 notrace unsigned long interrupt_exit_user_prepare(struct pt_regs *regs)
 {
-	unsigned long ret;
+	unsigned long ret = 0;
 
 	BUG_ON(regs_is_unrecoverable(regs));
 	BUG_ON(regs_irqs_disabled(regs));
@@ -193,7 +193,7 @@ notrace unsigned long interrupt_exit_user_prepare(struct pt_regs *regs)
 	kuap_assert_locked();
 
 	/* Clear exit_flags so only flags set during this exit are visible */
-	current_thread_info()->exit_flags = 0;
+	clear_thread_local_flags(_TLF_SYSCALL_EXIT_RESTOREALL);
 
 	local_irq_disable();
 again:
@@ -208,7 +208,8 @@ notrace unsigned long interrupt_exit_user_prepare(struct pt_regs *regs)
 
 	/* Restore user access locks last */
 	kuap_user_restore(regs);
-	ret = current_thread_info()->exit_flags & _TIF_RESTOREALL;
+	if (test_thread_local_flags(_TLF_SYSCALL_EXIT_RESTOREALL))
+		ret = _TIF_RESTOREALL;
 #ifdef CONFIG_PPC64
 	regs->exit_result = ret;
 #endif
diff --git a/arch/powerpc/kernel/ptrace/ptrace.c b/arch/powerpc/kernel/ptrace/ptrace.c
index 6cd180bc36ab2..c42839236cc85 100644
--- a/arch/powerpc/kernel/ptrace/ptrace.c
+++ b/arch/powerpc/kernel/ptrace/ptrace.c
@@ -23,6 +23,8 @@
 
 #include "ptrace-decl.h"
 
+struct user_pt_regs * upr;
+
 /*
  * Called by kernel/ptrace.c when detaching..
  *
@@ -119,7 +121,7 @@ long arch_ptrace(struct task_struct *child, long request,
 	case PTRACE_GETREGS:	/* Get all pt_regs from the child. */
 		return copy_regset_to_user(child, &user_ppc_native_view,
 					   REGSET_GPR,
-					   0, sizeof(struct user_pt_regs),
+					   0, sizeof(struct user_pt_regs) - sizeof(upr->__pt_regs_pad),
 					   datavp);
 
 #ifdef CONFIG_PPC64
@@ -128,7 +130,7 @@ long arch_ptrace(struct task_struct *child, long request,
 	case PTRACE_SETREGS:	/* Set all gp regs in the child. */
 		return copy_regset_from_user(child, &user_ppc_native_view,
 					     REGSET_GPR,
-					     0, sizeof(struct user_pt_regs),
+					     0, sizeof(struct user_pt_regs) - sizeof(upr->__pt_regs_pad),
 					     datavp);
 
 	case PTRACE_GETFPREGS: /* Get the child FPU state (FPR0...31 + FPSCR) */
@@ -293,13 +295,13 @@ void __init pt_regs_check(void)
 	CHECK_REG(PT_RESULT, result);
 	#undef CHECK_REG
 
-	BUILD_BUG_ON(PT_REGS_COUNT != sizeof(struct user_pt_regs) / sizeof(unsigned long));
+	BUILD_BUG_ON(PT_REGS_COUNT != (sizeof(struct user_pt_regs) - sizeof(upr->__pt_regs_pad)) / sizeof(unsigned long));
 
 	/*
 	 * PT_DSCR isn't a real reg, but it's important that it doesn't overlap the
 	 * real registers.
 	 */
-	BUILD_BUG_ON(PT_DSCR < sizeof(struct user_pt_regs) / sizeof(unsigned long));
+	BUILD_BUG_ON(PT_DSCR < (sizeof(struct user_pt_regs) - sizeof(upr->__pt_regs_pad)) / sizeof(unsigned long));
 
 	// ptrace_get/put_fpr() rely on PPC32 and VSX being incompatible
 	BUILD_BUG_ON(IS_ENABLED(CONFIG_PPC32) && IS_ENABLED(CONFIG_VSX));
diff --git a/arch/powerpc/kernel/signal.c b/arch/powerpc/kernel/signal.c
--- a/arch/powerpc/kernel/signal.c
+++ b/arch/powerpc/kernel/signal.c
@@ -356,6 +356,6 @@ void signal_fault(struct task_struct *tsk, struct pt_regs *regs,
 void arch_do_signal_or_restart(struct pt_regs *regs)
 {
 	BUG_ON(regs != current->thread.regs);
-	current_thread_info()->exit_flags |= _TIF_RESTOREALL;
+	set_thread_local_flags(_TLF_SYSCALL_EXIT_RESTOREALL);
 	do_signal(current);
 }
